### PR TITLE
Fix program can't be set to any

### PR DESCRIPTION
**Pull request by PfurtschellerP**
If ApplicationName of an existing rule should be set to any the parameter has to be set to $null which is not allowed. In these cases Set-NetFirewallRule can be used to set the parameter Program to "any" which translates to setting the ApplicationName of the rule to $null.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

win_firewall_rule

##### ADDITIONAL INFORMATION

https://github.com/ansible-collections/community.windows/issues/471

```paste below

```

### CODE REVIEW

File: plugins/modules/win_firewall_rule.ps1 [hunk 1]
```diff
@@ -166,7 +166,11 @@ try {
     if ($null -ne $enabled) { $new_rule.Enabled = $enabled } else { $new_rule.Enabled = $true }
     if ($null -ne $description) { $new_rule.Description = $description }
     if ($null -ne $group) { $new_rule.Grouping = $group }
-    if ($null -ne $program -and $program -ne "any") { $new_rule.ApplicationName = [System.Environment]::ExpandEnvironmentVariables($program) }
+    if ($null -ne $program -and $program -ne "any") { 
+        $new_rule.ApplicationName = [System.Environment]::ExpandEnvironmentVariables($program) 
+    } elseif ($program -eq "any") {
+        $new_rule.ApplicationName = $program
```
Comment (jborean93) [inline plugins/modules/win_firewall_rule.ps1:172]: This means that when someone sets `program: any` and a new rule is created the rule will have the program set `any` and not all like it does today. Where you trying to fix something by changing this logic?

**Review by jborean93 — CHANGES_REQUESTED** [review]
I'm reluctant in bringin in `Set-NetFirewallRule`, especially since the powershell cmdlets don't have a way of getting a firewall rule by ID. This means if there are multiple rules with the same name they will all have the `Program` set rather than the one we are working on. In saying that I tried out the code and found that yes it fails to unset the ApplicationName.

Digging a bit deeper it seems like this is a fundamental problem with the COM binder in PowerShell and how it treats null values. Luckily there is a workaround, you can add the following to the top of the module

```powershell
#AnsibleRequires -PowerShell Ansible.ModuleUtils.AddType
```

Then after https://github.com/ansible-collections/community.windows/blob/f9157edc9ea202f4dd558153cf652a330cf72fcf/plugins/modules/win_firewall_rule.ps1#L131 add the following

```powershell
$_remote_tmp = Get-AnsibleParam $params "_ansible_remote_tmp" -type "path" -default $env:TMP
Add-CSharpType -TempPath $_remote_tmp -References @'
using System;
using System.Runtime.InteropServices;

namespace Community.Windows.WinFirewallRule
{
    [ComImport]
    [InterfaceType(ComInterfaceType.InterfaceIsIDispatch)]
    [Guid("AF230D27-BABA-4E42-ACED-F524F22CFCE2")]
    public interface INetFwRule
    {
        [DispId(3)]
        string ApplicationName { get; set; }
    }

    public static class NetFwRule
    {
        public static void PutApplicationName(object rule, string value)
        {
            ((INetFwRule)rule).ApplicationName = string.IsNullOrEmpty(value) ? null : value;
        }
    }
}
'@
```

Now when the ApplicationName needs to be set it can do it with

```powershell
ElseIf(($prop -eq 'ApplicationName') {
    $value = if ($new_rule.$prop -eq "any") { $null } else { $new_rule.$prop }
    [Community.Windows.WinFirewallRule.NetFwRule]::PutApplicationName($existingRule, $value)
}
```

It would also be great if you could add a test for this scenario to https://github.com/ansible-collections/community.windows/tree/f9157edc9ea202f4dd558153cf652a330cf72fcf/tests/integration/targets/win_firewall_rule.

File: plugins/modules/win_firewall_rule.ps1 (hunk 1 above)
Comment (PfurtschellerP) [reply plugins/modules/win_firewall_rule.ps1:172]: Yes, `Set-NetFirewallRule` accepts `any` for the program parameter and nulls the property. I won't be neccessary anymore when applying your solution.

**Comment by PfurtschellerP** [comment]
@jborean93 is this now correct?

**Review by jborean93 — APPROVED** [review]
Yes thank you very much for implementing the new changes and writing the test for this specific scenario.

**Pull request merged**